File: quick_sort/quick_sort.py
#QuickSort

def quick_sort(items, left, right):
    if(len(items) > 1):
        index = partition(items, left, right)
        if(left < index - 1):
            quick_sort(items, left, index - 1)
        if(index < right):
            quick_sort(items, index, right)
    return items

def partition(items, left, right):
    pivot = items[(right + left)/2]
    while(left <= right):
        while(items[left] < pivot and left <= right):
            left += 1
        while(items[right] > pivot and right >= left):
            right -= 1
        if(left <= right):
            swap(items, left, right)
            left += 1
            right -= 1
    return left

def swap(items, lindex, rindex):
    temp = items[lindex]
    items[lindex] = items[rindex]
    items[rindex] = temp
    return None

items = [5, 4, 7, 5, 1, 6]
print items
quick_sort(items, 0, len(items) - 1)
print items

File: sortingChallenges/quick_sort/quickSort.js
function quickSort(items, left, right) {
    if(left === undefined){
      left = 0;
    }
    if(right === undefined){
      right = items.length - 1;
    }
    var index;
    if (items.length > 1) {
        index = partition(items, left, right);
        if (left < index - 1) {
            quickSort(items, left, index - 1);
        }
        if (index < right) {
            quickSort(items, index, right);
        }
    }
    return items;
}
function partition(items, left, right) {
    var pivot   = items[Math.floor((right + left) / 2)],
        i       = left,
        j       = right;
    while (i <= j) {
        while (items[i] < pivot) {
            i++;
        }
        while (items[j] > pivot) {
            j--;
        }
        if (i <= j) {
            swap(items, i, j);
            i++;
            j--;
        }
    }
    return i;
}
function swap(items, firstIndex, secondIndex){
    var temp = items[firstIndex];
    items[firstIndex] = items[secondIndex];
    items[secondIndex] = temp;
}

var items = [4, 2, 6, 1, 3, 9]
console.log("Original: ", items)
console.log("Sorted: ", quickSort(items));